kayenta-stackdriver/src/main/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsService.java

```diff
@@ -399,10 +399,22 @@ public List<MetricSet> queryMetrics(
               : stackdriverCanaryScope.getEnd();
 
       // TODO(duftler): What if there are no data points?
-      List<Double> pointValues =
-          points.stream()
+      List<Double> pointValues;
```

Comment:
I don't see any tests for this class, so it's a bit of a leap to make the first one.  Would you mind though?  It would help to have tests to verify this.  Thanks!

Reply:
I would love to, but I haven't developed anything serious in Java in over a decade. Would you be willing to provide guidance? Either by pointing me to documentation or chatting directly?

Reply:
Hmmm, this could be tough.  Ideally there would already be some mention of StackdriverMetricsService is some test class (*/src/test/<something>), but I don't see one.  There is [StackdriverCanaryMetricSetQueryConfigSpec.groovy](https://github.com/spinnaker/kayenta/blob/00b404a56687092c99dacd2eb3be1d316d396976/kayenta-stackdriver/src/test/groovy/com/netflix/kayenta/providers/metrics/StackdriverCanaryMetricSetQueryConfigSpec.groovy) but I don't know enough to know if that could exercise this functionality.

The most direct thing would be to make a new file `kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java` but then comes the task of finding an example to follow, and I'm coming up a bit empty.  Maybe [WavefrontMetricsServiceTest](https://github.com/spinnaker/kayenta/blob/master/kayenta-wavefront/src/test/java/com/netflix/kayenta/wavefront/metrics/WavefrontMetricsServiceTest.java) provides some reasonable inspiration?

In theory you could run these tests by executing:
```
$ ./gradlew kayenta-stackdriver:test --tests StackdriverMetricsServiceTest
```

I'm also happy to chat.  Are you on [spinnaker slack](https://spinnaker.io/docs/community/#slackhttpsjoinslackcomtspinnakerteamshared_invitezt-7juwxmx0-nq4ud4pjcbupykx3sxwqrg) ?

Reply:
Prometheus has solid integration tests.  The in memory tests are good, as are the signal fx tests.  See the SignalFX EndToEndCanaryIntegrationTests that test the integration completely as a good test case.  It's probably more overkill for this particular use case, but would welcome that.  

Barring that level of integration, you can extract out the code for some of the calls to test the response.  This is another set of code that could use some refactoring to cleanup how it operates.  The queryMetrics implementation is just a bit long for my tastes...

Reply:
I managed to create the file `kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java` and to create a dummy test:

```java
  @Test
  public void testDummy() {
    assertThat("dummy", is("dummy"));
  }
```

At this point I need to think how and what to mock such that `queryMetrics` will read a `timeSeries` that I can create in the test.

What I found is that `queryMetrics` obtains a `timeSeries` list like so:

```java
Monitoring.Projects.TimeSeries.List list =
    monitoring
        .projects()
        .timeSeries()
        .list("projects/" + stackdriverCredentials.getProject())
        .setAggregationAlignmentPeriod(alignmentPeriodSec + "s")
        .setAggregationCrossSeriesReducer(crossSeriesReducer)
        .setAggregationPerSeriesAligner(perSeriesAligner)
        .setFilter(filter)
        .setIntervalStartTime(stackdriverCanaryScope.getStart().toString())
        .setIntervalEndTime(stackdriverCanaryScope.getEnd().toString());
```
and the monitoring object is from the stackdriver API and created like so:

```java
import com.netflix.kayenta.security.AccountCredentialsRepository;

//...

    GoogleNamedAccountCredentials stackdriverCredentials =
        accountCredentialsRepository.getRequiredOne(metricsAccountName);
    Monitoring monitoring = stackdriverCredentials.getMonitoring();
```

Any tips on how to continue from here? I'm thinking how can I inject a mock of `monitoring` that I can control in the test, but I don't see a way how I can inject it directly.